### PR TITLE
Add discovery headers and first support for TEC REST API

### PR DESCRIPTION
**Pull request by lucatume**
Ticket: https://central.tri.be/issues/70712

This PR:

* reorders and consolidates the tests
* adds a rest test suite
* adds support for TEC REST API discovery headers in pretty and ugly permalink mode

Note:

* our compat is WP `3.9` which means no WP REST API support whatsoever; due to this I'm using WP `4.4` REST API functions if available and putting some needed ones in place if not    
* at this stage no REST API endpoint is really available yet; in another PR I will add an endpoint to answer on `/wp-json/tec/v1/events` that will register and use the WP REST API if available (WP `4.4+`) or use the `do_parse_request` hook on lower versions.
* supporting the two options and maintaining them is an overhead that can be dealt with easily in code after the request routing phase and that will buy us wider support and a form of independence from where the WP REST API might go

See schema:
2017-01-02 at 12 40

### CODE REVIEW

**Comment by borkweb** [comment]
Hrm. I only see changes for:

> * reorders and consolidates the tests
> * adds a rest test suite

(which look good so far) And not:

> * adds support for TEC REST API discovery headers in pretty and ugly permalink mode

Am I missing something obvious?

**Comment by zbtirrell** [comment]
I was just thinking the same. 

The diagram above looks perfect.

**Comment by zbtirrell** [comment]
I'd also like to get a feel for how much extra effort it is to support versions prior to 4.4. If it is a lot, I'd be fine with saying this feature/functionality requires 4.4+

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:x: | String concat is not required here; use a single string instead | [Main.php:57](https://github.com/moderntribe/the-events-calendar/blob/302921a5f68d1e63357ae2b0a9d2df9a3b0d5245/src/Tribe/RESTv1/Main.php#L57)

> via jenkins/codesniffer-phpcs

**Comment by lucatume** [comment]
> I'd also like to get a feel for how much extra effort it is to support versions prior to 4.4. If it is a lot, I'd be fine with saying this feature/functionality requires 4.4+

In coding terms it means conditionally adding a controller on the `do_parse_request` filter or registering a rewrite rule (slower) to handle requests on the REST API endpoint (`/wp-json/tec/v1`) with the same handler that's handling requests when WP 4.4 is available.  
There would be some massaging to make to provide the handler always the same request format but, for the time being, it would amount to what we allow the people to set filters for.

Good news is we can leave support for < 4.4 behind for the time being and work on 4.4+ only; if there are millions requesting it then we can implement it.

**Comment by lucatume** [comment]
I had not pushed files... silly me

**Comment by lucatume** [comment]
Adding to the answer above I'd leave the header when requesting import from a < 4.4 site to allow us to reply in a constructive way to distinguish between:

* not a TEC site
* yes a TEC site but with WP < 4.4 -> "Ask the site owner to upgrade her WordPress installation" and TEC version to tap into new and amazing TEC powers!"

The latter could be done returning something like `na` in the `X-TEC-API-VERSION` header

**Comment by lucatume** [comment]
On the same line we could return `disabled` if the user deactivated the REST API from Events > Settings (currently not in this PR)

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:x: | There must be a single space after the first parenthesis | [Main.php:115](https://github.com/moderntribe/the-events-calendar/blob/da53e8b689a2c263153d2fbe66d56d594c888be8/src/Tribe/RESTv1/Main.php#L115)
:x: | There must be a single space before the last parenthesis | [Main.php:115](https://github.com/moderntribe/the-events-calendar/blob/da53e8b689a2c263153d2fbe66d56d594c888be8/src/Tribe/RESTv1/Main.php#L115)
:x: | Each line in an array declaration must end in a comma | [Settings.php:57](https://github.com/moderntribe/the-events-calendar/blob/da53e8b689a2c263153d2fbe66d56d594c888be8/src/Tribe/RESTv1/Settings.php#L57)
:x: | Expected 1 space after &quot;=&gt;&quot;; 2 found | [Settings.php:65](https://github.com/moderntribe/the-events-calendar/blob/da53e8b689a2c263153d2fbe66d56d594c888be8/src/Tribe/RESTv1/Settings.php#L65)
:x: | empty() and isset() must only contain variables (PHP 5.2 compatibility) | [System.php:41](https://github.com/moderntribe/the-events-calendar/blob/da53e8b689a2c263153d2fbe66d56d594c888be8/src/Tribe/RESTv1/System.php#L41)

> via jenkins/codesniffer-phpcs

**Comment by lucatume** [comment]
Ok, this is ready for code review. Keep in mind this is **not adding** the endpoints yet but just a sensible and optionable discovery mode to our REST API.

File: src/Tribe/RESTv1/Main.php [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+
+	/**
+	 * Binds the implementations needed to support the REST API.
+	 */
+	public function bind_implementations() {
+		tribe_singleton( 'tec.rest-v1.headers', 'Tribe__Events__RESTv1__Headers__Supported' );
+		tribe_singleton( 'tec.rest-v1.settings', 'Tribe__Events__RESTv1__Settings' );
+		tribe_singleton( 'tec.rest-v1.system', 'Tribe__Events__RESTv1__System' );
+
+		include_once ( dirname( dirname( dirname( __FILE__ ) ) ) ) . '/functions/advanced-functions/rest-v1.php';
```
Comment (borkweb) [inline src/Tribe/RESTv1/Main.php]: Maybe:

```php
include_once Tribe__Events__Main::instance()->plugin_path . 'src/functions/advanced-functions/rest-v1.php`;
```

File: src/functions/advanced-functions/rest-v1.php [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+	 *
+	 * @global WP_Rewrite $wp_rewrite
+	 *
+	 * @param int         $blog_id Optional. Blog ID. Default of null returns URL for current blog.
+	 * @param string      $path    Optional. TEC REST route. Default '/'.
+	 * @param string      $scheme  Optional. Sanitization scheme. Default 'rest'.
+	 *
+	 * @return string Full URL to the endpoint.
+	 */
+	function tribe_rest_url( $blog_id = null, $path = '/', $scheme = 'rest' ) {
```
Comment (borkweb) [inline src/functions/advanced-functions/rest-v1.php]: Multisite instances probably won't be the norm.  I wonder if there is merit having a separate function that doesn't require `$blog_id` or perhaps re-ordering the parameters so passing `null` as the first param isn't the standard use case?

**Comment by lucatume** [comment]
Both good ideas and  both in place; ready for another pass.

**Comment by lucatume** [comment]
After a chat with Gustavo I'm reworking this a bit to move stuff to `common` and keep the door open for future APIs; will update when done.

**Comment by lucatume** [comment]
After some refactoring here it is (see https://github.com/moderntribe/tribe-common/pull/308 first)

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:x: | There must be a single space after the first parenthesis | [Main.php:27](https://github.com/moderntribe/the-events-calendar/blob/f3c196c6833ab9f8066d1dc07334212204f7c1cd/src/Tribe/REST/V1/Main.php#L27)
:x: | There must be a single space before the last parenthesis | [Main.php:27](https://github.com/moderntribe/the-events-calendar/blob/f3c196c6833ab9f8066d1dc07334212204f7c1cd/src/Tribe/REST/V1/Main.php#L27)
:x: | There must be a single space after the first parenthesis | [Main.php:43](https://github.com/moderntribe/the-events-calendar/blob/f3c196c6833ab9f8066d1dc07334212204f7c1cd/src/Tribe/REST/V1/Main.php#L43)
:x: | There must be a single space before the last parenthesis | [Main.php:43](https://github.com/moderntribe/the-events-calendar/blob/f3c196c6833ab9f8066d1dc07334212204f7c1cd/src/Tribe/REST/V1/Main.php#L43)

> via jenkins/codesniffer-phpcs

**Comment by borkweb** [comment]
When you get the spacing issues that tr1b0t complained about taken care of, this will be looking great. I'm happy to see some stuff move to common!

**Comment by borkweb** [comment]
I'm going to add `hold` on this to make sure we don't merge it until after 4.4 is released.

**Comment by lucatume** [comment]
rebasing this pull request against the `feature/rest-api` branch to keep the work in one place

**Pull request merged**